### PR TITLE
ExcludeforNtime

### PR DESCRIPTION
**Pull request by costr**
Users can exclude members for N time or through X date.

/pick exclude @JoeBob -For 5 days
/pick exclude @JoeBob -For 2 weeks
/pick exclude @JoeBob -Until 2/29/2020

for fun
/pick exclude @JoeBob -For 1 eon

You should also be able to do a time based exclude and then a separate general exclude (two commands)
/pick exclude @JoeBob -For 1 week
/pick exclude @BillyBob
You should see a message showing both are excluded but JoeBob will have a date through which he'll be excluded.

I also added some validation handling to respond to the end-user.

### CODE REVIEW

File: Picker/Func/Helpers/Settings/ChannelSettingsBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            var returnMembers = existingExcludedMembers.Where(x => !x.DateToExcludeTill.HasValue || x.DateToExcludeTill >= DateTime.Today).ToList();
+
+            if (membersToCleanup)
+            {
+#pragma warning disable 4014
+                Task.Run(async () =>
+                {
+                    await UpdateExcludeMembersSetting(returnMembers);
+                }).ConfigureAwait(false);
+#pragma warning restore 4014
```
Comment (osmyn) [inline Picker/Func/Helpers/Settings/ChannelSettingsBuilder.cs]: I'm curious why not just to _ = UpdateExcludeMembersSetting(returnMembers) here? I can't remember exactly what ConfigureAwait(false) does; maybe you are not firing-and-forgetting?

File: Picker/Func/Helpers/Settings/ExcludeSettingsBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+                case "weeks":
+                    return number * 7;
+                case "month":
+                case "months":
+                    throw new Exception("Only days and weeks are supported for member exclusion. Submit feedback if you'd like to request a differnt time frame.");
+                case "year":
+                case "years":
+                    throw new Exception("Only days and weeks are supported for member exclusion. Submit feedback if you'd like to request a differnt time frame.");
+                case "eons":
+                    throw new Exception("All we can do is try to shut it down, shut it down"); // 311 easter egg
```
Comment (osmyn) [inline Picker/Func/Helpers/Settings/ExcludeSettingsBuilder.cs]: :heart:

File: Picker/Func/IntentHandler/SettingsSlashHandler.cs [hunk 3]
```diff
@@ -8,60 +8,77 @@
...
+                {
+                    case SlackIntent.ExcludeMember:
+                    case SlackIntent.IncludeMember:
+                        await ProcessExcludeSettings(slack, intent);
+                        break;
+                }
+            }
+            catch (System.Exception ex)
+            {
+                // TODO: Maybe log what happened -- what are your thoughts?
```
Comment (osmyn) [inline Picker/Func/IntentHandler/SettingsSlashHandler.cs]: I think we should log it to our errors channel

Comment (osmyn) [reply Picker/Func/IntentHandler/SettingsSlashHandler.cs]: ooh, maybe not. I think we should log an unexpected error; maybe we could throw custom validation errors and not log those, then leave this one generic exception for if something wacky happened

File: Picker/Func/Helpers/Settings/ChannelSettingsBuilder.cs (hunk 1 above)
Comment (costr) [reply Picker/Func/Helpers/Settings/ChannelSettingsBuilder.cs]: My aim here was to fire and forget. Don't wait for the result just fire the clean-up in the background and keep moving forward with the response.

File: Picker/Func/IntentHandler/SettingsSlashHandler.cs (hunk 3 above)
Comment (costr) [reply Picker/Func/IntentHandler/SettingsSlashHandler.cs]: Ok...I'll make a ValidationException and implement that 👍

**Pull request merged**